Update to allow alternative Blocks to handle Grouping interactions

getdave opened the pull request:
This PR will allow alternative Blocks to be used to handle Grouping interactions. Currently, you can only use `core/group` but this PR will make it possible to register any valid Block as the Block to use when Grouping.

Previously the method that determined whether a given Block was the block to be used for “Grouping” interactions was hardcoded as “core/group. This PR builds upon https://github.com/WordPress/gutenberg/pull/15774 to allow alternative Blocks to be used to handle the Grouping interactions.

This should address some of the concerns raised in https://github.com/WordPress/gutenberg/pull/14908#issuecomment-501534012, namely:

- [x] Avoid hardcoding refs to `core/group` within `packages/blocks/src/api/factory.js`
- [x] Allow other Blocks to be registered to handle Grouping if required

## Screenshots

Below shows me registering the `test/alternative-group-block` Block from a custom e2e testing Plugin as the default Grouping Block. I'm then using it within the Grouping interactions.

![Screen Capture on 2019-07-03 at 13-19-25](https://user-images.githubusercontent.com/444434/60590978-44b65a00-9d95-11e9-9f17-daa778927e0f.gif)


## How has this been tested?

* Unit tests
* e2e tests
* Manual testing using the method outlined below.

## How to test

* Open `wp-admin` in the Gutenberg docker env
* Activate the `Gutenberg Test Custom Grouping Block` Plugin (**important**: be sure to _deactivate_ this after you have finished testing!)
* Create a new Page and add some Blocks to be Grouped.
* In the browser console 
  * type `window.wp.blocks.getGroupingBlockName()` to get the current Grouping Block name - it should be `core/group` (the default).
  * type `wp.blocks.setGroupingBlockName( 'test/alternative-group-block' );` - it should register that block (provided by the Plugin activated above) as the default for Grouping
* using the Editor UI
  * "Group" the Blocks using the "Group" item from the dropdown menu
  * Ungroup the Blocks again
  * Group the Blocks using the "Transform" control by selecting the "Alternative Group Block"

## Questions

* ~~Is relying on `getGroupingBlockName()` from the package [here](https://github.com/WordPress/gutenberg/pull/16278/files#diff-2be030b99c851ec84dc58d21ae932a93R58) appropriate? I'm concerned I'm not using a `select()` here.~~ Now uses `select()`.
* Should we have some kind of validation in place before we allow custom blocks to be set as the Grouping Block? At the moment you can register any old Block but only Blocks which support the necessary transforms and utilise `InnerBlocks` are allowed. For example, you could _try_ and register `coblocks/column` to act as the Grouping Block (it supports `InnerBlocks` after all) but within the correct `convert()` (transform) definition it won't function as you'd expect. 
* Where should I write documentation to cover this feature (if at all)?


## Types of changes

New feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

getdave committed:
Updates factory util to reference registered Grouping Block

Previously the method that determined whether a given Block was the block to be used for “Grouping” interactions was hardcoded as “core/group” within `isContainerBlock`. Updated to utilise the registered Grouping Block in order to allow alternative Blocks to be utilised for Grouping.

getdave committed:
Attempt e2e test to validate usage of alternate Block for grouping

Currently this doesn’t work because window doesn’t have NODE defined.

getdave committed:
Adds e2e test to confirm ability to register alternative Block for Grouping

Now using custom Gutenberg Plugin. Currently failing as the test Block appears to be unavailable within the Editor and doesn’t pass the `canInsertBlockType` test.

gziolo commented on packages/blocks/src/api/test/factory.js:
```diff
@@ -1554,12 +1555,34 @@ describe( 'block factory', () => {
 	describe( 'isContainerGroupBlock', () => {
-		it( 'should return true when passed block name matches "core/group"', () => {
-			expect( isContainerGroupBlock( 'core/group' ) ).toBe( true );
+		beforeEach( () => {
```
So parent `describe` unregisters all blocks for each test. That was confusing at first but it's perfectly valid 👍

getdave replied:
Yep. I'm adding tests to an existing suite so didn't want to disrupt.

gziolo commented on packages/e2e-tests/specs/block-grouping.test.js:
```diff
@@ -173,4 +175,37 @@ describe( 'Block Grouping', () => {
 		} );
 	} );
+
+	describe( 'Registering alternative Blocks to handle Grouping interactions', () => {
```
This is exactly as I imagined when we discussed the possible solution on Slack (links requires registration):

https://wordpress.slack.com/archives/C02QB2JS7/p1561454889265000

getdave replied:
That's good news!

getdave commented:
Looks like `switchToBlockType` is returning `null`. This is causing problems with `replaceBlocks` which is expecting a block and instead is receiving `null`. Need to dive deeper...

getdave committed:
Utilise withSelect to get the current Grouping Block name

getdave committed:
Removes unrequired test suite

getdave committed:
Fix bug with misnamed variable

getdave commented:
@aduth @gziolo This PR is now passing all builds and is "ready to go" feature wise. However it still needs another tech review and manual test. Much appreciated.

aduth reviewed (commented):
Should this other reference be updated as well?

https://github.com/WordPress/gutenberg/blob/f51f66202e285f62e92628bbc241a5090485d19c/packages/block-editor/src/components/block-actions/index.js#L120-L121

Less important, comment updates:

https://github.com/WordPress/gutenberg/blob/f51f66202e285f62e92628bbc241a5090485d19c/packages/blocks/src/api/factory.js#L150

https://github.com/WordPress/gutenberg/blob/f51f66202e285f62e92628bbc241a5090485d19c/packages/blocks/src/api/factory.js#L377

https://github.com/WordPress/gutenberg/blob/f51f66202e285f62e92628bbc241a5090485d19c/packages/editor/src/components/convert-to-group-buttons/convert-button.js#L102

getdave committed:
Remove hardcoded group block ref

Addresses https://github.com/WordPress/gutenberg/pull/16278#pullrequestreview-258540930

getdave committed:
Remove / update refs to core/group within comments

getdave commented:
@aduth I've addressed [those CR points](https://github.com/WordPress/gutenberg/pull/16278#pullrequestreview-258540930).

aduth reviewed (approved):
Another potentially useful consequence of these functions is to better accommodate scenarios where the grouping block isn't available (either because it's not been registered or in allowing to explicitly unset by `setGroupingBlockName( null )`).

aduth committed:
Block API: Stub default attributes, keywords values for block type registration

aduth committed:
Revert "Block API: Stub default attributes, keywords values for block type registration"

This reverts commit be6e7679d0d30be5945da13700f573b6ecece0eb.

getdave commented:
Thanks @aduth 👍
